src/core/MIGRATION.md

```diff
@@ -0,0 +1,1187 @@
+  * Introduce new plugin definition shim
+  * Introduce application shim
+  * Switch to new platform services
+  * Migrate to the new plugin system
+* Frequently asked questions
+  * Is migrating a plugin an all-or-nothing thing?
+  * Do plugins need to be converted to TypeScript?
+  * How is static code shared between plugins?
+  * How is "common" code shared on both the client and server?
+  * When does code go into a plugin, core, or packages?
+
+Make no mistake, it is going to take a lot of work to move certain plugins to the new platform. Our target is to migrate the entire repo over to the new platform throughout 7.x and to remove the legacy plugin system no later than 8.0, and this is only possible if teams start on the effort now.
+
+The goal of this document is to guide teams through the recommended process of migrating at a high level. Every plugin is different, so teams should tweak this plan based on their unique requirements.
+
+We'll start with an overview of how plugins work in the new platform, and we'll end with a generic plan of action that can be applied to any plugin in the repo today.
+
+## Overview
+
+Plugins in the new platform are not especially novel or complicated to describe. Our intention wasn't to build some clever system that magically solved problems through abstractions and layers of obscurity, and we wanted to make sure plugins could continue to use most of the same technologies they use today, at least from a technical perspective.
+
+New platform plugins exist in the `src/plugins` and `x-pack/plugins` directories.
+
+### Architecture
+
+Plugins are defined as classes and exposed to the platform itself through a simple wrapper function. A plugin can have browser side code, server side code, or both. There is no architectural difference between a plugin in the browser and a plugin on the server, which is to say that in both places you describe your plugin similarly, and you interact with core and/or other plugins in the same way.
+
+The basic file structure of a new platform plugin named "demo" that had both client-side and server-side code would be:
+
+```
+src/plugins
+  demo
+    kibana.json [1]
+    public
+      index.ts [2]
+      plugin.ts [3]
+    server
+      index.ts [4]
+      plugin.ts [5]
+```
+
+**[1] `kibana.json`** is a static manifest file that is used to identify the plugin and to determine what kind of code the platform should execute from the plugin:
+
+```json
+{
+  "id": "demo",
+  "server": true,
+  "ui": true
+}
+```
+
+Note that `package.json` files are irrelevant to and ignored by the new platform.
+
+**[2] `public/index.ts`** is the entry point into the client-side code of this plugin. It must export a function named `plugin`, which will receive a standard set of core capabilities as an argument (e.g. logger). It should return an instance of its plugin definition for the platform to register at load time.
+
+```ts
+import { PluginInitializerContext } from '../../../core/public';
```

Comment:
There's a lot to absorb in this PR and it took me an inordinate amount of time to realize that the imports that wander up through the hierarchy are simply pulling in type information. I think it would be helpful to move the import location to the top level - `/core`, or include types in the module name - `../../../core/public/types`, or better yet both - `/core/types`.

The current patter obscures the point that a similar import of code isn't allowed.

Reply:
We talked a little about this outside of this PR, but I'll reiterate my thoughts here:

Depending on context, things like classes, types, and interfaces can be used interchangeably in a source file, and it usually doesn't matter to a consumer how a type definition was defined. If you attempt to use an interface like a class, you will get a compiler error, so while the distinction isn't clear at import time, it should not result in bugs.

On the other hand, if I'm the author of a module and I decide to switch an implementation detail of my module to use an instance of a class instead of a manually typed object, I would expect to be able to export that class in place of the original type definition without it being a breaking change.

Either way, this migration guide is attempting to document how things rather than deciding how things should be. Please open a discuss issue if you want to consider this proposal further.